numpy/f2py/cfuncs.py

```diff
@@ -1007,36 +1007,39 @@
     PyObject* tmp = NULL;
     if (PyFloat_Check(obj)) {
         *v = PyFloat_AsDouble(obj);
-        return !(*v == -1.0 && PyErr_Occurred());
+        if (*v == -1.0) { goto error; }
+        return 1;
     }
 
     tmp = PyNumber_Float(obj);
     if (tmp) {
         *v = PyFloat_AsDouble(tmp);
         Py_DECREF(tmp);
-        return !(*v == -1.0 && PyErr_Occurred());
+        if (*v == -1.0){ goto error; }
+        return 1;
     }
 
     if (PyComplex_Check(obj)) {
-        PyErr_Clear();
         tmp = PyObject_GetAttrString(obj,\"real\");
     }
     else if (PyBytes_Check(obj) || PyUnicode_Check(obj)) {
         /*pass*/;
     }
     else if (PySequence_Check(obj)) {
-        PyErr_Clear();
         tmp = PySequence_GetItem(obj, 0);
     }
 
     if (tmp) {
+        PyErr_Clear(); // clear errors before calling again
```

Comment:
This is the only place the errors shouldn't be returned; i.e. when the function is calling itself.